File: sdks/go/dcp/v2/principal_mirror.go
// Package v2 — DCP-09 Principal Mirror (Go port).

package v2

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
)

func computeAuditChainHash(entries []map[string]interface{}) (string, error) {
	h := sha256.New()
	for _, entry := range entries {
		canonical, err := CanonicalizeV2(entry)
		if err != nil {
			return "", err
		}
		h.Write([]byte(canonical))
	}
	return "sha256:" + hex.EncodeToString(h.Sum(nil)), nil
}

// MirrorParams bundles inputs for GenerateMirror.
type MirrorParams struct {
	MirrorID        string
	SessionNonce    string
	AgentID         string
	HumanID         string
	Period          map[string]string // {"from": "...", "to": "..."}
	AuditEntries    []map[string]interface{}
	Narrative       string
	DecisionSummary string
}

// GenerateMirror produces a signed principal-mirror artefact over a narrative
// and an audit-chain integrity hash.
func GenerateMirror(
	registry *AlgorithmRegistry,
	classicalKey CompositeKeyInfo,
	pqKey CompositeKeyInfo,
	p MirrorParams,
) (map[string]interface{}, error) {
	auditHash, err := computeAuditChainHash(p.AuditEntries)
	if err != nil {
		return nil, err
	}
	payload := map[string]interface{}{
		"dcp_version":         "2.0",
		"mirror_id":           p.MirrorID,
		"session_nonce":       p.SessionNonce,
		"agent_id":            p.AgentID,
		"human_id":            p.HumanID,
		"period":              p.Period,
		"narrative":           p.Narrative,
		"action_count":        len(p.AuditEntries),
		"decision_summary":    p.DecisionSummary,
		"audit_chain_hash":    auditHash,
		"timestamp":           utcNowISO(),
	}
	canonical, err := CanonicalizeV2(payload)
	if err != nil {
		return nil, fmt.Errorf("canonicalize: %w", err)
	}
	sig, err := CompositeSign(registry, CtxDelegation, []byte(canonical), classicalKey, pqKey)
	if err != nil {
		return nil, err
	}
	payload["composite_sig"] = sig
	return payload, nil
}

File: sdks/python/dcp_ai/v2/principal_mirror.py
"""
DCP-09 v2.0 Principal Mirror — Python port.

Generates human-readable narrative summaries of agent actions synthesised
from audit chains. The narrative itself is produced by the caller; this
module signs it and binds it to an audit-chain integrity hash.
"""

from __future__ import annotations

import hashlib
from typing import Any

from dcp_ai.v2.canonicalize import canonicalize_v2
from dcp_ai.v2.composite_ops import CompositeKeyInfo, composite_sign
from dcp_ai.v2.crypto_registry import AlgorithmRegistry
from dcp_ai.v2.domain_separation import DCP_CONTEXTS
from dcp_ai.v2.lifecycle import _utc_now


def _compute_audit_chain_hash(entries: list[dict[str, Any]]) -> str:
    """Hash the audit entries in order for integrity binding."""
    h = hashlib.sha256()
    for entry in entries:
        h.update(canonicalize_v2(entry).encode("utf-8"))
    return "sha256:" + h.hexdigest()


async def generate_mirror(
    registry: AlgorithmRegistry,
    classical_key: CompositeKeyInfo,
    pq_key: CompositeKeyInfo,
    *,
    mirror_id: str,
    session_nonce: str,
    agent_id: str,
    human_id: str,
    period: dict[str, str],
    audit_entries: list[dict[str, Any]],
    narrative: str,
    decision_summary: str,
) -> dict[str, Any]:
    """Produce a signed principal-mirror artifact over a narrative + audit trail."""
    audit_chain_hash = _compute_audit_chain_hash(audit_entries)

    payload: dict[str, Any] = {
        "dcp_version": "2.0",
        "mirror_id": mirror_id,
        "session_nonce": session_nonce,
        "agent_id": agent_id,
        "human_id": human_id,
        "period": period,
        "narrative": narrative,
        "action_count": len(audit_entries),
        "decision_summary": decision_summary,
        "audit_chain_hash": audit_chain_hash,
        "timestamp": _utc_now(),
    }
    canonical = canonicalize_v2(payload)
    composite = await composite_sign(
        registry, DCP_CONTEXTS["Delegation"], canonical.encode("utf-8"), classical_key, pq_key
    )
    return {**payload, "composite_sig": composite}
